fix: matmul shape for lazy

yangguohao opened the pull request:
Fix the shape of lazy after matmul
```
from autoray.lazy import Variable
import numpy as np
a = Variable(shape=(4, 3))
b = Variable(shape=(3, ))
print(a@b)
# <LazyArray(fn=matmul, shape=(4, 3), backend='None')>

a = np.random.rand(4, 3)
b = np.random.rand(3)
print((a@b).shape)
# (4,)
```

jcmgray commented:
Thanks for the PR! Just looking at the [`numpy.matmul`](https://numpy.org/doc/stable/reference/generated/numpy.matmul.html) docstring, it says:



    If both arguments are 2-D they are multiplied like conventional matrices.

    If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.

    If the first argument is 1-D, it is promoted to a matrix by prepending a 1 to its dimensions. After matrix multiplication the prepended 1 is removed.

    If the second argument is 1-D, it is promoted to a matrix by appending a 1 to its dimensions. After matrix multiplication the appended 1 is removed.

If possible could we handle all of these cases, and also add some tests?

yangguohao commented:
Sure. I will start working on it.

jcmgray commented:
Great, thanks for the updates and contribution!

yangguohao commented:
Hi Johnnie, I have added code to handle different cases and make sure it raise exceptions for dimension mismatches to facilitate code inspection, otherwise it is difficult to debug when it comes to actual computations. Test cases were also added.
